src/rt/ds/hashmap.h

```diff
@@ -256,411 +195,272 @@ namespace verona::rt
+      auto it = Iterator(*this, 0);
+      if (unmark_key(key_of(slots[0])) == 0)
+        ++it;
 
-        if (dib_entry == dib_other)
-        {
-          // This entry is already present. This should only happen for the
-          // original o, not for any swapped pointer.
-          if (
-            (key_of(entry) == orig_key) &&
-            (key_of(entry) == unmark_key(other_key)))
-          {
-            location = index;
-            return false;
-          }
-        }
-        else if (dib_entry > dib_other)
-        {
-          auto tmp = std::move(other);
+      return it;
+    }
 
-          if (key_of(entry) == orig_key)
-            location = index;
+    Iterator end() const
+    {
+      return Iterator(*this, capacity());
+    }
 
-          // The DIB of the entry to insert is greater than the DIB of the
-          // entry at this index. Insert o here, and continue looking for
-          // somewhere to insert other.
-          set_entry(index, std::forward<E>(entry), dib_entry);
+    /**
+     * Find an entry in the map with the given key and return an iterator to the
+     * corresponding entry. If no entry exitsts, the return value will be equal
+     * to the return value of `end()`.
+     */
+    Iterator find(const Object* key) const
+    {
+      if (key == nullptr)
+        return end();
 
-          key_of(tmp) = (uintptr_t)get_pointer(key_of(tmp));
-          entry = std::move(tmp);
-          dib_entry = dib_other;
-        }
+      const auto hash = bits::hash((void*)key);
+      auto index = hash % capacity();
+      for (size_t probe_len = 0; probe_len <= longest_probe; probe_len++)
+      {
+        const auto k = unmark_key(key_of(slots[index]));
+        if (k == (uintptr_t)key)
+          return Iterator(*this, index);
 
-        // Advance both the index, wrapping around, and the DIB.
-        index = (index + 1) & mask;
-        dib_entry++;
+        if (++index == capacity())
+          index = 0;
       }
 
-      assert(0);
-      return false;
-    }
-
-    template<typename E>
-    bool insert(Alloc* alloc, E entry)
-    {
-      size_t dummy;
-      return insert(alloc, std::forward<E>(entry), dummy);
+      return end();
     }
 
+    /**
+     * Insert an entry into the map. The first element of the returned pair will
+     * be true if a new key is inserted, and false if an existing entry is
+     * updated. The second element of the returned pair is an iterator to the
+     * inserted entry. The key of the inserted entry must not be null.
+     */
     template<typename E>
-    void insert_unique(Alloc* alloc, E entry)
-    {
-      auto unique = insert(alloc, std::forward<E>(entry));
-      assert(unique);
-      UNUSED(unique);
-    }
-
-    void sweep_set(Alloc* alloc, size_t marked)
+    std::pair<bool, Iterator> insert(Alloc* alloc, E entry)
     {
-      if (size_bits == 0)
-        return;
+      // Resize at ~.8 load factor.
+      if (size() >= ((capacity() * 4) / 5))
```

Comment:
Why resize here?  and not just when you hit a long probe.  This branch is on a fast path, overflowing the probe length, is pulling the conditional later.

Reply:
You're right, the tests almost always reach above 0.93 load factor before needing to resize.